xcube/core/store/preload.py

```diff
@@ -0,0 +1,409 @@
+        """Exit the context. Calls ``close()``."""
+        self.close()
+
+
+class NullPreloadHandle(PreloadHandle):
+    """Null-pattern implementation of a ``PreloadHandle``."""
+
+    def get_state(self, data_id: str) -> PreloadState:
+        return PreloadState(data_id)
+
+    @property
+    def cancelled(self) -> bool:
+        return False
+
+    def cancel(self):
+        pass
+
+    def show(self) -> Any:
+        return None
+
+
+class ExecutorPreloadHandle(PreloadHandle):
+    """TODO - Add docs"""
+
+    def __init__(
+        self,
+        data_ids: tuple[str, ...],
+        preload_data: Callable[[PreloadHandle, str], None] | None = None,
+        executor: Executor | None = None,
+        blocking: bool = True,
+    ):
+        self._preload_data = preload_data
+        self._executor = executor or ThreadPoolExecutor()
+        self._blocking = blocking
+
+        self._states = {data_id: PreloadState(data_id=data_id) for data_id in data_ids}
+        self._cancel_event = threading.Event()
+        self._display = PreloadDisplay.create(list(self._states.values()))
+        self._lock = threading.Lock()
+        self._futures: dict[str, Future[str]] = {}
+        for data_id in data_ids:
+            future: Future[str] = self._executor.submit(self._run_preload_data, data_id)
+            future.add_done_callback(self._handle_preload_data_done)
+            self._futures[data_id] = future
+
+        if blocking:
+            self._display.show()
+            self._executor.shutdown(wait=True)
+
+    def get_state(self, data_id: str) -> PreloadState:
+        return self._states[data_id]
+
+    @property
+    def cancelled(self) -> bool:
+        """Return true if and only if the internal flag is true."""
+        return self._cancel_event.is_set()
+
+    def cancel(self):
+        self._cancel_event.set()
+        for future in self._futures.values():
+            future.cancel()
+        self._executor.shutdown(wait=False)
+
+    def notify(self, event: PreloadState):
+        state = self._states[event.data_id]
+        if (
+            event.status is not None
+            and event.status != state.status
+            and state.status
+            in (PreloadStatus.stopped, PreloadStatus.cancelled, PreloadStatus.failed)
+        ):
+            # Status cannot be changed
+            return
+        with self._lock:
+            state.update(event)
+            self._display.update()
+
+    def preload_data(self, data_id: str):
+        """Preload the data resource given by *data_id*.
+
+        Concurrently executes the *preload_data* passed to the constructor,
+        if any. Otherwise, it does nothing.
+
+        Can be overridden by clients to implement the actual preload operation.
+
+        Args:
+           data_id: The data identifier of the data resource to be preloaded.
+        """
+        if self._preload_data is not None:
+            self._preload_data(self, data_id)
```

Comment:
I am trying to understand why we have this method here. Seems redundant to me. In the method below `_run_preload_data`, we can directly call the preload method passed in.
So I have the following doubt:
- The role of this class is to handle the preload jobs such as updating the status, progress, message etc., cancelling the preload event and showing the user the overall status. Then why do we need this method here as it seems we are mixing responsibilities? The preload_data() function (not this method) that we pass in as a dependency should be responsible for how the data is preloaded and this class should be oblivious to it. I would be interested in knowing about the intuition behind having this method here.

Reply:
See the docstring:

> Can be overridden by clients to implement the actual preload operation.

Reply:
The idea is to support 2 use cases:

- derive from `ExecutorPreloadHandle` and override `preload_data()` method
- use `ExecutorPreloadHandle` as-is and pass your own `preload_data` function